woocommerce/woocommerce-ios
# Add dateFromString helper function to TestKit

Closes: #2551

### Description

Problem: We use the same (or very similar) DateFormatter function across dozens of our Unit Tests, this is generally added as an extension to each individual file. This function is used to convert from String to Date type, for example in [ProductPriceSettingsValidatorTests.swift](https://github.com/woocommerce/woocommerce-ios/blob/762d30f9b166741c99f64a5d74009fba68f67d3d/WooCommerce/WooCommerceTests/ViewRelated/Products/Edit%20Product/Edit%20Price/ProductPriceSettingsValidatorTests.swift#L147-L149)

The function can be found [here](https://github.com/woocommerce/woocommerce-ios/blob/81ed7b18078c58d3337f53a91d0f09c802bf3c3f/Networking/Networking/Extensions/DateFormatter%2BWoo.swift#L14):

```
        /// Date And Time Formatter
        ///
        public static let dateTimeFormatter: DateFormatter = {
            let formatter = DateFormatter()
            formatter.locale = Locale(identifier: "en_US_POSIX")
            formatter.timeZone = TimeZone(identifier: "GMT")
            formatter.dateFormat = "yyyy'-'MM'-'dd'T'HH:mm:ss"
            return formatter
        }()
```

Proposed solution: This PR creates a shared helper function that can be used across unit tests. As these belong to different frameworks, like WooCommerce, Yosemite, or Networking, we can declare this shared helper in `TestKit`, where we already have general assert functions and helpers for unit tests.

The new helper function also allows optional parameters for the locale, timezone, and date format. Or defaults to the old ones if nothing needs to be changed. Thanks @itsmeichigo for the idea!

### Testing instructions

As there are a bunch of files that needed changes, I made different commits grouping these files by Framework to make it easier.

1 - Run the tests
2 - Note that the Test Suite is executed and passes successfully.

### Todo:

- [x] Check if all Unit Tests pass
- [x] Change the rest of the Unit Tests where this function is used.
- [x] Investigate: Do we need to hardcode these dates on tests? Or we can create them in the Helper? For example some sort of `date.random()`

---
